docs: add a tag-matching design document

phlogistonjohn opened the pull request:
### What does this PR achieve? Why do we need it?

This is a design doc for a proposed enhancement to heketi for a tag-matching mechanism allows user specified filtering of devices to implement tiering, etc.

raghavendra-talur commented:
@phlogistonjohn This is awesome! To be able to get this feature based on many underlying parts that were done for Arbiter and Zone matching feature is great.

raghavendra-talur commented on docs/design/tag-matching.md:
````diff
@@ -0,0 +1,72 @@
+make use of the tagging:
+```
+    $ heketi-cli volume create --size=5 --gluster-volume-options 'user.heketi.tag-match tier=gold'
+    $ heketi-cli volume create --size=100 --gluster-volume-options 'user.heketi.tag-match tier=silver'
````
I am thinking we need to have device in the key for example `user.heketi.device-tag-match` because we might later want to match only some nodes with more CPU etc(we already have for zone).

phlogistonjohn replied:
OK, sounds good to me.

raghavendra-talur commented:
@phlogistonjohn Also, I would like to think of a way to protect some devices(like ssd) from being used in initial deployment.

For example:
gk-deploy OR openshift-ansible would provide the topology file and tag devices. It would also immediately ask for a volume for heketidbstorage. Admin might not want the default volumes to use "gold" tier. 

We have arbiter only tags for such use case in arbiter feature. I think we need something similar here too.

phlogistonjohn commented:
> @phlogistonjohn Also, I would like to think of a way to protect some devices(like ssd) from being used in initial deployment.
> 
> For example:
> gk-deploy OR openshift-ansible would provide the topology file and tag devices. It would also immediately ask for a volume for heketidbstorage. Admin might not want the default volumes to use "gold" tier.
> 
> We have arbiter only tags for such use case in arbiter feature. I think we need something similar here too.

Heketi today has the ability to specify the tags in the topology file. Combine that with recent work to support creating heketidbstorage with volume options and I think this is mainly verifying the pieces work together correctly. I really do not want to create a special case for things like 'heketidbstorage' even though they are common because this feature should be more about mechanism than policy IMO. I'd rather work with those projects to make sure they adapt to new heketi features than 2nd guess what they are doing.

phlogistonjohn commented:
I think instead of specifying a special tag, we just require users who don't want certain volumes used by initial setup processes to esablish a pattern of tags like so:
The software such as gk-deploy runs `./heketi-cli  setup-openshift-heketi-storage --gluster-volume-options 'user.heketi.tag-match builtinvols!=disabled'`
Then the user is requied to tag devices as builtinvols:disabled in the topology file ahead of time so that the set up process will skip over such devices.

raghavendra-talur commented:
I think I did not explain well. Here is what I intended

# Cases where some devices are to be used only when required
If admin wants to have some special devices like ssd that ought to be used only for certain volumes, it can be achieved by adding the user.heketi.tag-match key with a value that does not match such devices in  `pre_request_volume_options` in heketi.json.

phlogistonjohn commented:
I've changed the name of the volume option to "user.heketi.device-tag-match"

I've added sections to cover considerations related to but outside the immediate deign of the feature as we discussed.

phlogistonjohn commented:
OK. It actually helps to push the change one is talking about.

raghavendra-talur commented:
@phlogistonjohn Thanks for update. The design document looks good to me.
